src/server/scripts/EasternKingdoms/Karazhan/boss_chess_event.cpp

```diff
@@ -0,0 +1,2081 @@
+                }
+
+                break;
+            }
+        }
+
+        Talk(TALK_MEDIHV_CHEAT_EMOTE);
+    }
+
+    Creature* GetPiece(Creature* sourcePiece, uint8 searchType, float distance, bool hostile, uint32 minHPDiff, bool checkFront)
+    {
+        if (!sourcePiece)
+        {
+            return nullptr;
+        }
+
+        Creature* target = nullptr;
+        std::vector<Creature*> targets;
+
+        for (uint8 row = 0; row < MAX_ROW; ++row)
+        {
+            for (uint8 col = 0; col < MAX_COL; ++col)
+            {
+                if (Creature* piece = ObjectAccessor::GetCreature(*me, _boards[row][col].pieceGUID))
+                {
+                    if (IsFriendly(piece, sourcePiece) == hostile)
+                    {
+                        continue;
+                    }
+
+                    if (checkFront && !sourcePiece->HasInArc(float(M_PI) / 3.0f, piece))
+                    {
+                        continue;
+                    }
+
+                    if (minHPDiff)
+                    {
+                        if (piece->GetMaxHealth() - piece->GetHealth() <= minHPDiff)
+                        {
+                            continue;
+                        }
+
+                        minHPDiff = piece->GetMaxHealth() - piece->GetHealth();
+                    }
+
+                    float dist = sourcePiece->GetExactDist2d(piece);
+                    if (dist < distance)
+                    {
+                        if (searchType == CHESS_PIECE_SEARCH_TYPE_CLOSEST)
+                        {
+                            distance = dist;
+                            target = piece;
+                        }
+                        else
+                        {
+                            targets.push_back(piece);
+                        }
+                    }
+                }
+            }
+        }
+
+        if (!targets.empty())
+        {
+            target = Acore::Containers::SelectRandomContainerElement(targets);
+        }
+
+        return target;
+    }
+
+    void UpdateAI(uint32 diff) override
+    {
+        uint32 chessPhase = _instance->GetData(DATA_CHESS_GAME_PHASE);
+        if (chessPhase != CHESS_PHASE_INPROGRESS_PVE && chessPhase != CHESS_PHASE_INPROGRESS_PVP)
+            return;
+
+        if (chessPhase == CHESS_PHASE_INPROGRESS_PVE)
+        {
+            if (_cheatTimer <= diff)
+            {
+                HandleCheat();
+                _cheatTimer = urand(45000, 100000);
+            }
+            else
+            {
+                _cheatTimer -= diff;
+            }
+        }
+    }
+
+    void sGossipHello(Player* player) override
+    {
+        uint32 chessPhase = _instance->GetData(DATA_CHESS_GAME_PHASE);
+        switch (chessPhase)
+        {
+            case CHESS_PHASE_NOT_STARTED:
+                AddGossipItemFor(player, GOSSIP_ICON_CHAT, "We want to play a game against you!", GOSSIP_SENDER_MAIN, MEDIVH_GOSSIP_START_PVE);
```

Comment:
After the repair server is restarted, the event should be PVP
```
if (_instance->GetBossState(DATA_CHESS_EVENT) == DONE)
    AddGossipItemFor(player, GOSSIP_ICON_CHAT, "PvP", GOSSIP_SENDER_MAIN, MEDIVH_GOSSIP_START_PVP); //We'd like to fight each other
else
    AddGossipItemFor(player, GOSSIP_ICON_CHAT, "We want to play a game against you!", GOSSIP_SENDER_MAIN, MEDIVH_GOSSIP_START_PVE);
```

Reply:
Repair server - what does it mean?
No need to check boss state.

Reply:
Don't we need to save it in 'instance'?

Reply:
My English is very bad. You can try it as I say:
1. "We want to play a game against you!"
2. Chess end
3. "PvP"
4. "Restart"
5. You will find that it appears again "We want to play a game against you!"
6. Can be loots again